Fix ztest mock parameter count issue

jettr opened the pull request:
If you tried to use more mock parameters than was set in the CONFIG_ZTEST_PARAMETER_COUNT, you would get an "unrelated" memory access violation in your test somewhere.

This PR
 - Fixes the out of bound array access
 - Cleans up comments and error strings
 - Cleans up prototype usage of ztest_expect_data to not required caller to cast to `void *`
 - Increases the default `CONFIG_ZTEST_PARAMETER_COUNT` from 1 to 10 when mocking is enabled (still guarded by `CONFIG_ZTEST_MOCKING`)

Commit:
ztest: increase default mock parameter count

If one enables mocking, then they most likely want more than 1 mockable
or verifiable parameter by default. Bump the default number of mockable
parameters to 10 to handle most cases without needing to touch an
additional Kconfig option.

Signed-off-by: Jett Rink <jettrink@google.com>

Commit:
ztest: fix failure when exceeding mock parameter count

If you tried to use more mock parameters than were set in
CONFIG_ZTEST_PARAMETER_COUNT, you would get a random seg fault somewhere
unrelated to the mocking parameter as we would use memory past an array
bounds.

Signed-off-by: Jett Rink <jettrink@google.com>

sjg20 reviewed (commented):
commit-message comments

Should use present tense.

Also your first paragraph describes the problem, so should add a sentence explaining what the commit does to fix it.

sjg20 commented on subsys/testsuite/ztest/include/ztest_mock.h:
```diff
@@ -55,33 +55,33 @@
- * Only data pointer is stored by this function so it must still be valid when
- * ztest_check_expected_data is called.
+ * When using ztest_check_expected_data(), the data pointed to by
+ * @a param should be same @a data in this function. Only data pointer is stored
```
I think these changes should be in a separate commit

sjg20 replied:
(I mean your bug fix should be one commit and all your clean-ups in another)

jettr replied:
Done

sjg20 reviewed (changes requested):
commit-message comments

Should use present tense.

Also your first paragraph describes the problem, so should add a sentence explaining what the commit does to fix it.

Commit:
ztest: clean up comments, strings, and interface

Clean up comment typos and wording for expect data functions.

Add a cast to (void *) to match convention set in expect value;
this allows the caller to avoid casting on every single call.

Signed-off-by: Jett Rink <jettrink@google.com>

jettr commented:
From what I can see on the build failures, it looks like sanitycheck just stopped executing without any failures: https://buildkite.com/zephyr/zephyr/builds/4565#5ee1c14a-6653-4a56-b04e-0f7cdb732aea

Is there a way to kick the build again? Is this kind of failure expected/transient? There is still a possibility that there is something wrong with these patches, but they do work for me locally (and there isn't much too them either).

jettr commented:
All checks are passing. @nashif do you merge this or is there some automated process that will merge this since it has all of the necessary approvals and passing checks?
